snuba/web/rpc/common.py

```diff
@@ -0,0 +1,281 @@
+    ):
+        if context is None or len(context.project_ids_to_names) == 0:
+            raise BadSnubaRPCRequestException(
+                "If you request project_name, the AttributeKeyTransformContext must be "
+                "sent. Snuba doesn't have access to project names directly."
+            )
+        return f.transform(
+            column("project_id"),
+            literals_array(
+                None, [literal(k) for k in context.project_ids_to_names.keys()]
+            ),
+            literals_array(
+                None, [literal(v) for v in context.project_ids_to_names.values()]
+            ),
+            literal("unknown"),
+            alias="project_name",
+        )
+
+    if attr_key.name == "trace_id" and attr_key.type == AttributeKey.Type.TYPE_STRING:
+        return f.CAST(column("trace_id"), "String", alias="trace_id")
+
+    if attr_key.name in HEX_ID_COLUMNS:
+        if attr_key.type == AttributeKey.Type.TYPE_STRING:
+            return f.hex(column(attr_key.name))
+        raise BadSnubaRPCRequestException(
+            f"Attribute {attr_key.name} must be requested as a string, got {attr_key.type}"
+        )
+
+    if attr_key.name in TIMESTAMP_COLUMNS:
+        if attr_key.type == AttributeKey.Type.TYPE_STRING:
+            return f.CAST(column(attr_key.name), "String", alias=attr_key.name)
+        if attr_key.type == AttributeKey.Type.TYPE_INT:
+            return f.CAST(column(attr_key.name), "Int64", alias=attr_key.name)
+        if attr_key.type == AttributeKey.Type.TYPE_FLOAT:
+            return f.CAST(column(attr_key.name), "Float64", alias=attr_key.name)
+        raise BadSnubaRPCRequestException(
+            f"Attribute {attr_key.name} must be requested as a string, float, or integer, got {attr_key.type}"
+        )
+
+    if attr_key.name in NORMALIZED_COLUMNS:
+        if NORMALIZED_COLUMNS[attr_key.name] == attr_key.type:
+            return column(attr_key.name)
+        raise BadSnubaRPCRequestException(
+            f"Attribute {attr_key.name} must be requested as {NORMALIZED_COLUMNS[attr_key.name]}, got {attr_key.type}"
+        )
+
+    # End of special handling, just send to the appropriate bucket
+    if attr_key.type == AttributeKey.Type.TYPE_STRING:
+        return SubscriptableReference(
+            alias=attr_key.name, column=column("attr_str"), key=literal(attr_key.name)
+        )
+    if attr_key.type == AttributeKey.Type.TYPE_FLOAT:
+        return SubscriptableReference(
+            alias=attr_key.name, column=column("attr_num"), key=literal(attr_key.name)
+        )
+    if attr_key.type == AttributeKey.Type.TYPE_INT:
+        return f.CAST(
+            SubscriptableReference(
+                alias=attr_key.name,
+                column=column("attr_num"),
+                key=literal(attr_key.name),
+            ),
+            "Int64",
+        )
+    if attr_key.type == AttributeKey.Type.TYPE_BOOLEAN:
+        return f.CAST(
```

Comment:
I'm seeing a lot of casts in this function, why are they necessary?

Reply:
if you request a boolean, this returns a boolean? it'd be weird if you requested a boolean and got a number back